### DISCUSSION TITLE
What is the exit code once sanboot fail

### DISCUSSION THREAD

**Discussion by amdhugo** [ipxe/ipxe#278] [Q&A] [answered]
I see the demonstration on https://ipxe.org/cmd/sanboot#command_status
Failure Booting was unsuccessful

I wonder what the exit code is once booting was unsuccessful? Will ipxe forward the control to BIOS then?

**Comment by mcb30** [comment, marked as answer]
For BIOS, any failure would be reported as https://ipxe.org/err/0b8080 since the operating system on the SAN drive can only indicate "failure" (with no specified reason).

For UEFI, the error would be whatever the operating system bootloader had returned as an error.

In both cases, control will return to iPXE.  What happens next will depend entirely on what you have asked iPXE to do.  For example, your script could choose to drop to a shell if booting failed:
```
sanboot iscsi:192.168.0.1::::iqn.2010-04.org.ipxe:demo || shell
```

If you have not asked iPXE to do anything explicit, then the default action will be for the script to exit (since any unhandled error will always cause a script to exit).

**Reply by amdhugo** [reply]
I hope to exit, but it looks like ipxe crash or itself exit and not execute next step. 
The command in default profile is like,
sanboot --drive 0x81
:next menu1
chain xxxxx
:next menu2
chain xxxxx

The interesting thing is BIOS behavior is good after sanboot failure but if replace sanboot with 'exit' or 'exit -1', BIOS doesn't work as expected. So I hope to capture the return code from sanboot --driver 0x81

BTW, how --driver 0x81 doesn't work always? Isn't it boot into a Harddisk?

**Reply by mcb30** [reply]
> I hope to exit, but it looks like ipxe crash or itself exit and not execute next step.
> The command in default profile is like,
> sanboot --drive 0x81

As I said: If you have not asked iPXE to do anything explicit, then the default action will be for the script to exit (since any unhandled error will always cause a script to exit).

See https://ipxe.org/scripting#error_handling for a further repetition of this statement. :slightly_smiling_face:

**Reply by amdhugo** [reply]
Refer to the description of 'Error handling', I think my question shall be what is the return code once iPXE terminate a script immediately if the line 'sanboot --driver 0x80' of my script fails?

**Reply by NiKiZe** [reply]
There is no error code from iPXE, or rather there is no concepts of error code.

You can only control what happens inside iPXE, outside is mostly undefined, not due to iPXE, but due to that BIOS has no dinition of what it should do.

**Comment by amdhugo** [comment]
[AMD Official Use Only - Internal Distribution Only]

Okay!
Is there a way to print out return code like ‘echo $?’?

**Reply by mcb30** [reply]
Within iPXE, you can use the `${errno}` variable to access the most recent error code.
